app/src/main/java/de/xikolo/services/DownloadService.kt

```diff
@@ -0,0 +1,310 @@
+        }
+
+    override fun onCreate() {
+        instance = this
+
+        if (Config.DEBUG) Log.d(TAG, "DownloadService created")
+
+        jobCounter = AtomicInteger()
+
+        notificationUtil = NotificationUtil(this)
+
+        // Don't use HttpLoggingInterceptor, crashes with OutOfMemoryException!
+        val client = OkHttpClient.Builder()
+            .addInterceptor(ApiService.authenticationInterceptor)
+            .addInterceptor(ApiService.userAgentInterceptor)
+            .build()
+        downloadClient = DownloadManager.Builder()
+            .context(this)
+            .downloader(OkHttpDownloader.create(client))
+            .build()
+
+        downloadMap = ConcurrentHashMap()
+
+        // Start up the thread running the service. Note that we create a
+        // separate thread because the service normally runs in the process's
+        // main thread, which we don't want to block.  We also make it
+        // background priority so CPU-intensive work will not disrupt our UI.
+        val thread = HandlerThread("ServiceStartArguments", Process.THREAD_PRIORITY_BACKGROUND)
+        thread.start()
+
+        // Get the HandlerThread's Looper and use it for our Handler
+        val serviceLooper = thread.looper
+        serviceHandler = ServiceHandler(serviceLooper)
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        if (Config.DEBUG) Log.d(TAG, "DownloadService start command received")
+
+        intent?.extras?.getString(ARG_TITLE)?.let { title ->
+            if (intent.extras?.getBoolean(ARG_SHOW_NOTIFICATION, true) == true) {
+                val titles = runningDownloadTitles
+                titles.add(title)
+
+                val notification = notificationUtil?.getDownloadRunningNotification(titles)?.build()
+                startForeground(NotificationUtil.DOWNLOAD_RUNNING_NOTIFICATION_ID, notification)
+            }
+
+            // For each start request, send a message to start a job and deliver the
+            // start ID so we know which request we're stopping when we finish the job
+            val msg = serviceHandler?.obtainMessage()
+            msg?.data = intent.extras
+            jobCounter?.incrementAndGet()
+            serviceHandler?.sendMessage(msg)
+        }
+
+        // If we get killed, after returning from here, restart
+        return Service.START_STICKY
+    }
+
+    @Synchronized
+    private fun stopJob() {
+        if (jobCounter?.decrementAndGet() == 0) {
+            stopSelf()
+        }
+    }
+
+    override fun onBind(intent: Intent): IBinder? {
+        // We don't provide binding, so return null
+        return null
+    }
+
+    override fun onDestroy() {
+        instance = null
+
+        if (Config.DEBUG) Log.d(TAG, "DownloadService destroyed")
+
+        stopForeground(true)
+        downloadClient?.cancelAll()
+        notificationUtil = null
+    }
+
+    @Synchronized
+    fun isDownloading(url: String): Boolean {
```

Comment:
`url` was actually meant be a nullable argument as it is safe to use because of the `getDownload()` which accepts a `null` argument

Reply:
Oh, okay. I changed that, but please double check.